packages/nimble-components/src/table/specs/table-row-selection.md

```diff
@@ -0,0 +1,136 @@
+# Table Row Selection HLD
+
+## Problem Statement
+
+The nimble-table should support selecting rows. Row selection should be configurable such that a client can enable single-row selection, multi-row selection, or no selection. A client of the table should be able to get the current selection from the nimble-table element and should also be notified when the selection changes.
+
+## Out of scope of this HLD
+
+Any implication that data hierarchy has on row selection is out of scope for this HLD.
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#856 Selection support in the table](https://github.com/ni/nimble/issues/856)
+-   [Table README](./README.md)
+-   [Table Design Doc](https://xd.adobe.com/view/5b476816-dad1-4671-b20a-efe796631c72-0e14/screen/d389dc1e-da4f-4a63-957b-f8b3cc9591b4/specs/)
+-   [Table interactions UX Spec](https://xd.adobe.com/view/6f3be15d-8503-4f1f-54b9-5bc27955b3e4-190a/)
+
+## Implementation / Design
+
+### API
+
+There will be a new attribute on the table named `selection-mode`. It will be of type `TableRowSelectionMode` as shown below, and it will default to `TableRowSelectionMode.none`.
+
+```ts
+export const TableRowSelectionMode = {
+    none: undefined,
+    single: 'single',
+    multiple: 'multiple'
+} as const;
+export type TableRowSelectionMode =
+    (typeof TableRowSelectionMode)[keyof typeof TableRowSelectionMode];
+```
+
+When row selection changes due to a user interaction, the table will emit a `selection-change` event. The event will not be emitted if selection changes due to a non-user action, such as updating the data. The `detail` of the `selection-change` event will be of type `TableRowSelectionEventDetail`, which will be defined as follows:
+
+```ts
+export interface TableRowSelectionEventDetail {
+    selectedRecordIds: string[];
+}
+```
+
+In the case where grouping is enabled, it is possible that a grouping row can be visualized as selected because all of its child rows are selected. It is important to note that because the grouping row does not have a record associated with it, it will never be returned as part of the `selectedRecordIds` array in the event detail. Instead, each of its selected child rows will have their record ID included in the `selectedRecordIds` array.
```

Comment:
I guess this also means that there's no 1-line way to programmatically change the selection of everything in a group at once. I think this is fine for now (and possibly forever), just wanted to note it.

Reply:
That is a good point, and I agree that it seems fine for now.

As an additional note, the [grouping HLD](https://github.com/ni/nimble/blob/caa23a5171750a2984e94b80ff6a280602910c34/packages/nimble-components/src/table/specs/table-row-grouping-hld.md) doesn't currently expose a concept of IDs on groups or grouped rows, so, right now, we don't have a programmatic way to reference one of those groups. We would need to add that concept first, and then it wouldn't be very difficult to add a programmatic way to select/deselect a group.